### PR TITLE
Initial pass at Metrics support in ILogger

### PR DESCRIPTION
**Pull request by anurse**
This implements #708. It's got some naive elements, like the way we handle filters (which involves some locking). I want to clean that up but wanted to get this into a PR first.

### CODE REVIEW

File: src/Microsoft.Extensions.Logging/Logger.cs [hunk 1]
```diff
@@ -136,6 +139,21 @@ public IDisposable BeginScope<TState>(TState state)
             return scope;
         }
 
+        public IMetric DefineMetric(string name)
+        {
+            if (!_metrics.TryGetValue(name, out var metric))
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging/Logger.cs]: Is dictionary safe for multi thread read during the write?

File: src/Microsoft.Extensions.Logging/Microsoft.Extensions.Logging.csproj [hunk 2]
```diff
@@ -13,6 +13,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" />
     <PackageReference Include="Microsoft.Extensions.Options" />
+    <PackageReference Include="System.Collections.Immutable" />
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging/Microsoft.Extensions.Logging.csproj]: Is it still used?

File: src/Microsoft.Extensions.Logging/Logger.cs [hunk 3]
```diff
@@ -197,5 +215,87 @@ public void Dispose()
...
+            private IMetric[] _metrics;
+            private object _sync = new object();
+
+            public Metric(Logger logger, string name)
+            {
+                _logger = logger;
+                _name = name;
+            }
+
+            public void RecordValue(double value)
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging/Logger.cs]: As we discussed this implementation has to be optimized to update metrics array only when required.

File: src/Microsoft.Extensions.Logging/Logger.cs [hunk 4]
```diff
@@ -197,5 +215,87 @@ public void Dispose()
...
+                    // REVIEW: LogLevel for metrics?
+                    // REVIEW: Filtering by Metric Name?
+                    if (_logger.Loggers[i].IsEnabled(LogLevel.Critical))
+                    {
+                        metrics[i].RecordValue(value);
+                    }
+                }
+            }
+
+            public void RecordValue<T>(double value, T properties) where T : IEnumerable<KeyValuePair<string, object>>
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging/Logger.cs]: Test for this and `TimeSpan` extension method.

File: src/Microsoft.Extensions.Logging/Logger.cs (hunk 1 above)
Comment (anurse) [reply src/Microsoft.Extensions.Logging/Logger.cs]: I'm honestly not sure. I was having the same concern and thinking I should just use ConcurrentDictionary

File: src/Microsoft.Extensions.Logging/Microsoft.Extensions.Logging.csproj (hunk 2 above)
Comment (anurse) [reply src/Microsoft.Extensions.Logging/Microsoft.Extensions.Logging.csproj]: Not anymore, will remove

**Comment by anurse** [comment]
Closing this, we're going to look at using [EventCounters](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventcounter?view=netcore-2.0) for our metrics needs.

**Pull request closed without merging**